### PR TITLE
Issue 49 - Compute cyclomatic complexity

### PR DESCRIPTION
**Pull request by racodond**
A basic implementation of https://github.com/SonarCommunity/sonar-css/issues/49
@kalidasya: Tell me what you think. Thanks!

### CODE REVIEW

**Comment by kalidasya** [comment]
I do not agree with it. A css file is more like the content of a function than a class itself. So having a at-rule or a selector is not a function, its more a function call (apply these rules on the result of these selectors). The complexity, (but I am not a css expert) is more coming from the compound selectors (as mentioned in the page), so your subscriptions for the complexity are ok, but it should be decreased by one for each selector and at-rule. 

```
mybox { /* 0 */
}

.mybox { /* 0 */
}

#mybox { /* 0 */
}

[target] { /* 0 */
}

:hover { /* 0 */
}

a:hover { /* +1 */
}

a.class:hover { /* +2 */
}

div a { /* +1 */
}

div a:hover span.myclass { /* +4 */
}

ul > div + a + li { /* + 3 */
}
```

but its is an important and central topic whatever we implement we have to make sure we have beta testers with multi language projects, to see how it influence their overall metrics.

**Comment by racodond** [comment]
> The complexity, (but I am not a css expert) is more coming from the compound selectors (as mentioned in the page), so your subscriptions for the complexity are ok

Good. This is the most important point from my point of view.

> but it should be decreased by one for each selector and at-rule.

I do not agree. Most simple pieces of code have a cyclomatic complexity of 1. The following piece of code would have an average "complexity/function" of 0 with your way of counting. That doesn't make sense. It would have an average "complexity/function" of 1 with my way of counting. Meaning, the simplest piece of code, that is the case from my point of view.

```
mybox { /* 0 */
}

.mybox { /* 0 */
}

#mybox { /* 0 */
}
```

> A css file is more like the content of a function than a class itself. So having a at-rule or a selector is not a function, its more a function call (apply these rules on the result of these selectors). 

I agree that a CSS rule set is not really a function in the way functions are defined in Java for example. But mapping rule sets as functions provide a way to display the number of rule sets on the SonarQube web interface and most important to compute an average complexity per function (meaning average complexity per rule set). This complexity per function metric is kind of a nice indicator of the complexity of your CSS as it tells you globally whether you have over-specified selectors or not.

> but its is an important and central topic whatever we implement we have to make sure we have beta testers with multi language projects, to see how it influence their overall metrics.

Agree. I tested on our Java/JavaScript/CSS code base and it's not bad. The average complexity/function is a bit lower than the Java part but not that much. And usually CSS is only a small part of your project, so it doesn't have such a big impact on the overall measure.

What do you think?

**Comment by kalidasya** [comment]
Ok you are right, basically we need to add 1 complexity for all files.
What I am afraid of, is if a project has a strong gui, probably they have one big css file for network optimization. For example the css of http://www.girlsgogames.co.uk/ has 1035 selector (approximately) it is not a 1035 complexity file in my opinion. (especially they do not work on this file directly, it is generated),  so if there is a css file like this in a project, suddenly the method/complexity will be decreased by a lot (majority will have 1, and this was the issue why I had to remove this connection),and files/complexity increased (not sure by how much, because I guess there will not be too many css files, but still.  

What if we say each file is a function as well, each file has 1 complexity, and we only increase the complexity if there are more than one sub-selector/etc I think there is a way to implement our own complexity visitor, so we can handle this cleanly. I think this is closer to reality.

**Comment by racodond** [comment]
> Ok you are right, basically we need to add 1 complexity for all files.

You meant "Add 1 complexity for each rule set", right?

> What I am afraid of, is if a project has a strong gui, probably they have one big css file for network optimization. For example the css of http://www.girlsgogames.co.uk/ has 1035 selector (approximately) it is not a 1035 complexity file in my opinion.

It is very complex from my point of view. A 1,035 rule set file is unmaintainable because it's way too complex. The good practice is to have several CSS files specific to some domains/areas. When you build your application you can aggregate them and minify the aggregation for network optimization.

> (especially they do not work on this file directly, it is generated), so if there is a css file like this in a project, suddenly the method/complexity will be decreased by a lot (majority will have 1, and this was the issue why I had to remove this connection),and files/complexity increased (not sure by how much, because I guess there will not be too many css files, but still.

You don't analyze this generated file. You would analyze your maintainable smaller CSS files.

> What if we say each file is a function as well, each file has 1 complexity, and we only increase the complexity if there are more than one sub-selector/etc I think there is a way to implement our own complexity visitor, so we can handle this cleanly. I think this is closer to reality.

If you have one single CSS file of 1,035 simple rule sets in your project, you'll get a complexity of 1 for this file. That doesn't reflect at all the complexity and unmaintainability of this file...

What do you think of my first proposal after reading the above comments?

**Comment by kalidasya** [comment]
well usually strong gui app development works with sass or less, what we do not support yet, so they can only analyse the end result. 
Even if we say they maintain that css by hand, 1035 complexity is way too much compared to like: http://nemo.sonarqube.org/drilldown/measures/713523?metric=file_complexity
what has 1036 complexity/file. What I think css complexity if we calculate it like that is not 1-1 comparable with complexity of other languages. We can try this solution, but we need some beta testers to evaluate it. 
I am concerned we will significantly increase the overall complexity, and decrease the complexity/function with this, but we can try to figure out is it really the case. Not sure what is the best way to reach the css plugin users.

**Comment by racodond** [comment]
> well usually strong gui app development works with sass or less, what we do not support yet, so they can only analyse the end result. 

Supporting sass or less should be one of the next steps of the plugin.
Meanwhile, they would/should analyze the generated CSS file(s) in a separate SonarQube project and only care about complexity/function (this would be meaningful), not complexity/file (this would not be meaningful as the file is an aggregation of others)

> Even if we say they maintain that css by hand, 1035 complexity is way too much compared to like: http://nemo.sonarqube.org/drilldown/measures/713523?metric=file_complexity
> what has 1036 complexity/file.

Not from my point of view. Let's take the rule about Java class complexity: http://nemo.sonarqube.org/coding_rules#q=class%20too%20complex|languages=java. The threshold is 200. Applying the same threshold to CSS with the proposed way of computing complexity seems coherent to me. It means 200 simple rule sets or 100 rule sets with two selectors per file. I wouldn't allow bigger CSS files.

> What I think css complexity if we calculate it like that is not 1-1 comparable with complexity of other languages.

Sure but it should be meaningful for CSS guys and in about the same range. It's meaningful from my point of view:
- The higher the complexity/function, the higher the average number of selectors per rule set
- The higher the complexity/file, the higher the average number of rule sets per file

On our code base:
- 25,470 lines of CSS code out of a 320,000 line of cod Java/JavaScript/CSS application
- Complexity/function:
  - on CSS only: 3.1 
  - overall: 2.2
- Complexity/file:
  - on CSS only: 251
  - overall: 10.4

Complexity/function => no deviation, we know that our CSS is way to complex and that we should simplify it; so 3.1 vs 2.2 is fine by me.
Complexity/file: => deviation indeed, we know that we should split our CSS because they are unmaintainable, but I agree that complexity/file will always be higher for CSS than Java for example. The same when you've got a mix of COBOL and Java. Complexity/file is always very very high on COBOL compare to Java. 
For me, the meaningful indicator to follow is complexity/function. We don't follow complexity/file at my company but have the evolution of complexity/function in our quality gate.

> We can try this solution, but we need some beta testers to evaluate it. 
> I am concerned we will significantly increase the overall complexity, and decrease the complexity/function with this, but we can try to figure out is it really the case. Not sure what is the best way to reach the css plugin users.

Not much idea. We can launch a vote that last a bit longer.

**Comment by racodond** [comment]
What about asking some feedback on this thread of discussion on the SonarQube Google Group?

**Comment by kalidasya** [comment]
well only one thing left, we cannot expect the user to set up a different project in case of generated css.  That is really not user friendly, but I think we can easily have a plugin config and not set functions and complexity.  I am pretty sure with this implementation we will still reopen the original issue SONARCSS-18.

Having a long vote period does not mean anyone will test it. Asking feedback is a good idea.

**Comment by racodond** [comment]
@kalidasya: Complexity computation can now be switched off. I'll let your review and merge.

File: css-squid/src/main/java/org/sonar/css/CssAstScanner.java [hunk 1]
```diff
@@ -106,6 +106,18 @@ public static SourceFile scanSingleFile(File file, SquidAstVisitor<LexerlessGram
...
       .build());
 
+    if (conf.getComputeComplexity()) {
+      builder.withSquidAstVisitor(CounterVisitor.<LexerlessGrammar>builder()
+        .setMetricDef(CssMetric.COMPLEXITY)
+        .subscribeTo(CssGrammar.CLASS_SELECTOR, CssGrammar.ATTRIBUTE_SELECTOR, CssGrammar.TYPE_SELECTOR, CssGrammar.ID_SELECTOR, CssGrammar.PSEUDO, CssGrammar.AT_RULE)
+        .build());
+    }
+
+    builder.withSquidAstVisitor(CounterVisitor.<LexerlessGrammar>builder()
```
Comment (kalidasya) [inline css-squid/src/main/java/org/sonar/css/CssAstScanner.java]: this must be in the if, that was the original issue we had. a lot of 0 complexity function decreased the overall complexity

File: css-squid/src/test/java/org/sonar/css/CssAstScannerTest.java [hunk 2]
```diff
@@ -78,4 +78,16 @@ public void atRules() {
...
+    SourceFile file = CssAstScanner.scanSingleFile(new File("src/test/resources/metrics/metric.css"));
+    assertThat(file.getInt(CssMetric.FUNCTIONS)).isEqualTo(6);
+  }
+
+  @Test
+  public void complexity() {
+    SourceFile file = CssAstScanner.scanSingleFile(new File("src/test/resources/metrics/complexity.css"));
+    assertThat(file.getInt(CssMetric.COMPLEXITY)).isEqualTo(21);
+  }
+
```
Comment (kalidasya) [inline css-squid/src/test/java/org/sonar/css/CssAstScannerTest.java:93]: a test to check it is still ok if the complexity calculation is off?

File: README.md [hunk 3]
```diff
@@ -26,3 +26,23 @@ A sample project is available on GitHub: https://github.com/SonarSource/sonar-ex
...
+* Attribute selector
+* Type selector
+* Pseudo-class selector
+* At-rule
+
+### Complexity/function
+It computes the complexity/rule, meaning the average number of selectors per rule.
+
+It gives a measurement on how specific the selectors are.
+
```
Comment (kalidasya) [inline README.md:48]: note about the possibility to turn it off?

File: css-squid/src/main/java/org/sonar/css/CssAstScanner.java (hunk 1 above)
Comment (racodond) [inline css-squid/src/main/java/org/sonar/css/CssAstScanner.java]: What should be in the if statement exactly?

Comment (kalidasya) [inline css-squid/src/main/java/org/sonar/css/CssAstScanner.java]: the function metric calculation:

```
builder.withSquidAstVisitor(CounterVisitor.<LexerlessGrammar>builder()
.setMetricDef(CssMetric.FUNCTIONS)
.subscribeTo(CssGrammar.SELECTOR, CssGrammar.AT_RULE)
.build());
```

File: README.md (hunk 3 above)
Comment (racodond) [inline README.md:48]: Fixed in https://github.com/racodond/sonar-css/commit/2df7c1171ba70b8614c197642ac39df61e38df13

File: css-squid/src/main/java/org/sonar/css/CssAstScanner.java (hunk 1 above)
Comment (racodond) [inline css-squid/src/main/java/org/sonar/css/CssAstScanner.java]: Fixed in https://github.com/racodond/sonar-css/commit/d64ecb6776c780fda003889c0d63ec8bed6d77e8

File: css-squid/src/test/java/org/sonar/css/CssAstScannerTest.java (hunk 2 above)
Comment (racodond) [inline css-squid/src/test/java/org/sonar/css/CssAstScannerTest.java:93]: I don't really know if it's worth it but here it is: https://github.com/racodond/sonar-css/commit/d7b9fde20474587bcea599d266eaa33d905f988d

File: css-squid/src/main/java/org/sonar/css/CssAstScanner.java [hunk 4]
```diff
@@ -52,11 +53,18 @@ private CssAstScanner() {
   /**
    * Helper method for testing checks without having to deploy them on a Sonar instance.
    */
-  public static SourceFile scanSingleFile(File file, SquidAstVisitor<LexerlessGrammar>... visitors) {
+  public static SourceFile scanSingleFile(File file, CssConfiguration configuration, SquidAstVisitor<LexerlessGrammar>... visitors) {
```
Comment (kalidasya) [inline css-squid/src/main/java/org/sonar/css/CssAstScanner.java]: this method is for testing only, why dont we create another signature instead and we dont need to change all the tests.

**Comment by racodond** [comment]
@kalidasya : It should now be OK.

**Pull request merged**